DOCSP-8055 adds page about indexes (#44)
mongodb/docs-node: source/fundamentals/indexes.txt

Query Coverage and Performance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When you execute a query against MongoDB, your query can include three
parts:

- query criteria that specify field(s) and value(s) you are looking for,
- options that affect the query's execution (e.g. read concern),

- (optional) projection criteria to specify the fields MongoDB should
  return

When all the fields specified in the query criteria and projection of a
query are indexed, MongoDB returns results directly from the index
without scanning any documents or bringing documents into memory. Queries
that use indexes are more performant than queries that rely on
collection scans.

The MongoDB manual addresses both :manual:`query coverage
</core/query-optimization/#read-operations-covered-query>` and
:manual:`index intersection </core/index-intersection>` (using multiple
indexes to fulfil a query) in more detail.

Operational Considerations
~~~~~~~~~~~~~~~~~~~~~~~~~~

To improve query performance, you should build indexes on fields that
appear often in queries and for all operations that involve sorted
results. Each index that you add consumes disk space and memory
when active so you should track index

> memory and disk usage for capacity planning. Adding indexes also has a

Reviewer: ```suggestion
memory and disk usage for capacity planning. Adding too many indexes can have a
```

Reviewer: I'm unsure its worth being entirely accurate here. Technically, all indexes will have some performance impact for writes, but this is kind of a given. I think we should frame the message as "too many".

> negative performance impact for write operations. When designing your data model and choosing the indexes your application requires, you must weigh the benefits and impacts of each index that you add.

Reviewer 2: I personally feel like clarifying the "why" for the negative performance impact upon writes might be helpful here, especially for users who are reading about indexes for the first time/aren't very familiar with them. Maybe linking this to the idea that indexes are good for values that are rarely written/updated, and mostly read?

Reviewer 2: (especially since the write performance hit is small compared to the potential read performance gains, which can be multiple orders of magnitude)

:manual:`Indexing Strategies </applications/indexes>` and :manual:`Data Modeling and Indexes </core/data-model-operations/#data-model-indexes>` provide additional insight into index behaviors and considerations.

Index Types
-----------

MongoDB supports a number of different index types to support querying
your data. The following sections describe the most common index types
and provide sample code that creates each index type.

Single Field and Compound Indexes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Single field indexes are user-defined ascending or descending indexes on

Reviewer 2: linking to a definition of ascending or descending might help here -- otherwise I think we might be better off without it, since the direction doesn't matter in 99% of cases.

Reviewer 2: Even though you mention it below, it's still kind of confusing in my opinion because *why* an index needs a direction is unclear.

a single field of a document. 

The following example uses the ``createIndex()`` method to create an
ascending index on the ``title`` field in the ``movies`` collection in
the ``sample_mflix`` database.

.. literalinclude:: /code-snippets/indexes/single-field.js
   :language: js
   :start-after: begin-ex
   :end-before: end-ex

> Compound indexes are user-defined indexes on *multiple* fields. When you create a compound index, you specify the direction (i.e. ascending or descending) for each field in the index.

Reviewer 2: Something along the lines of "a compound index organizes values first by field1, then by field2, etc. etc." might be useful to add in here as well without getting too low-level.

The following example uses the ``createIndex()`` method to create a compound index on the ``type`` and ``genre`` fields in the ``movies`` collection in the ``sample_mflix`` database.

.. literalinclude:: /code-snippets/indexes/compound.js
   :language: js
   :start-after: begin-ex
   :end-before: end-ex

Multikey Indexes (Indexes on Array Fields)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Multikey indexes are user-defined ascending or descending indexes on
fields that contain an array value. You define a multikey index in the

> same fashion as you would a single and compound indexes: you do not 

Reviewer 2: "a single" -> single

> same fashion as you would a single and compound indexes: you do not 

Reviewer: ```suggestion
same fashion as you would for both single and compound indexes: you do not 
```

need to explicitly specify that an index is multikey.

The following example use the ``createIndex()`` method to create an ascending index on the ``cast`` field in the ``movies`` collection in the ``sample_mflix`` database.

.. literalinclude:: /code-snippets/indexes/multikey.js
   :language: js
   :start-after: begin-ex
   :end-before: end-ex

Multikey indexes behave differently from non-multikey indexes in terms of
query coverage, index bound computation, and sort behavior. For a full
explanation of multikey indexes, including a discussion of their behavior
and limitations, refer to the :manual:`Multikey Indexes page
</core/index-multikey>` in the MongoDB manual.

Text Indexes
~~~~~~~~~~~~

> :manual:`Text indexes </core/index-text>` support text search queries on

Reviewer 2: Could you link to the text search page of the crud guide here?

string content. ``text`` indexes can include any field whose value is a
string or an array of string elements. MongoDB supports text search for
various languages. You can specify the default language as an option when
creating the index.

The following example uses the ``createIndex()`` method to create a
``text`` index on the ``fullplot`` field in the ``movies`` collection in
the ``sample_mflix`` database and specifies the default language to be
``english``.

.. literalinclude:: /code-snippets/indexes/text.js
   :language: js
   :start-after: begin-ex
   :end-before: end-ex

For a full explanation of text search with MongoDB, refer to :manual:`Text Indexes </core/index-text>` in the MongoDB manual.

Geospatial Indexes
~~~~~~~~~~~~~~~~~~

> MongoDB supports queries of geospatial coordinate data using :manual:`2dsphere </core/2dsphere>` indexes. With a ``2dsphere`` index, you can query for inclusion, intersection, and proximity. The :manual:`Geospatial Queries page </geospatial-queries>` in the MongoDB manual provides a detailed explanation of geospatial query operations in MongoDB.

Reviewer 2: could you link to the geo page of the crud guide here?

> MongoDB supports queries of geospatial coordinate data using :manual:`2dsphere </core/2dsphere>` indexes. With a ``2dsphere`` index, you can query for inclusion, intersection, and proximity. The :manual:`Geospatial Queries page </geospatial-queries>` in the MongoDB manual provides a detailed explanation of geospatial query operations in MongoDB.

Reviewer 2: technically geospatial queries can use either ``2d`` **or** ``2dsphere`` indexes -- I feel like it might be a good idea to mention both, even if we only want to provide an example of one. (unfortunately both have distinct use cases, so 2dsphere does not actually replace 2d)

To create a ``2dsphere`` index, the field you are indexing must be a :manual:`GeoJSON object </reference/geojson>`, as in the following sample document from the ``theaters`` collection in the ``sample_mflix`` database:

.. code-block:: javascript

   {
      "_id" : ObjectId("59a47286cfa9a3a73e51e75c"),
      "theaterId" : 104,
      "location" : {
         "address" : {
            "street1" : "5000 W 147th St",
            "city" : "Hawthorne",
            "state" : "CA",
            "zipcode" : "90250"
         },

>          "geo" : {

Reviewer 2: For an example, I feel like we might be better off ditching all the fields besides "geo" or at least simplifying them. Right now I feel like this makes it seem like geoJSON objects have a lot more data than they really do -- maybe we can meet in the middle and more specifically call out the location.geo field before throwing users into this example?
